From 49abcb659da158ea4369a4a08fdcd5a5926dd235 Mon Sep 17 00:00:00 2001
From: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
Date: Tue, 12 Feb 2019 14:47:28 +0200
Subject: [PATCH 05/12] v4l2-renderer: Release dma-buf when attaching null
 buffer

When renderer is requested to attach a null buffer, e.g. via
wl_surface_attach with wl_buffer pointer set to null, it is
expected that the current buffer is released, effectively
dropping our reference on the attached dma buffer. Otherwise,
the reference is kept until the renderer is destroyed preventing
the dma buffer from being released earlier.

Signed-off-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>
---
 libweston/v4l2-renderer.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/libweston/v4l2-renderer.c b/libweston/v4l2-renderer.c
index 7641bed8..7eb96373 100644
--- a/libweston/v4l2-renderer.c
+++ b/libweston/v4l2-renderer.c
@@ -1664,6 +1664,12 @@ v4l2_renderer_attach(struct weston_surface *es, struct weston_buffer *buffer)
 			weston_buffer_reference(&vs->buffer_ref, NULL);
 			return;
 		}
+	} else {
+		// null buffer is a special case: current buffer needs to be
+		// released, so reference counter of the attached
+		// dma buffer is dropped from us now
+		v4l2_release_dmabuf(vs);
+		v4l2_release_kms_bo(vs);
 	}
 
 #ifdef V4L2_GL_FALLBACK_ENABLED
-- 
2.17.1

